r/R/transforms.R

```diff
@@ -812,3 +812,71 @@ multiply_cols <- function(mat, vec) {
   assert_true(length(vec) == ncol(mat))
   t(t(mat) * vec)
 }
+
+
+
+#################
+# Linear residual
+#################
+
+setClass("TransformLinearResidual", contains = "TransformedMatrix")
+setMethod("iterate_matrix", "TransformLinearResidual", function(x) {
+  res <- iterate_matrix(x@matrix)
+  
+  if (any(nrow(x@row_params) == 0, nrow(x@col_params) == 0)) {
+    return(res)
+  }
+  iterate_matrix_linear_residual_cpp(res, x@row_params, x@col_params)
+})
+setMethod("short_description", "TransformLinearResidual", function(x) {
+  # Return multiple lines, one for each transform active
+  res <- short_description(x@matrix)
+  if (any(nrow(x@row_params) == 0, nrow(x@col_params) == 0)) {
+    res <- c(res, sprintf("Linear regress out 0 variable"))
+    return(res)
+  }
+  res <- c(res, sprintf("Linear regress out %d variable(s)", nrow(x@row_params) - 1))
+  res
+})
+
+#' Linear Regression
+#'
+#' Calculate the residuals from a linear model fit. 
+#'
+#' @param mat IterableMatrix (raw counts)
+#' @param latent_data Extra data to regress out, should be cells x latent `data.frame`
+#' @return IterableMatrix
+#' @export
+linear_regress_out <- function(mat, latent_data = NULL) {
+  if (length(latent_data) == 0) {
+    return(mat)
+  }
+  qr <- calc_qr(mat = mat, latent_data = latent_data)
+  Q <- as.matrix(qr.Q(qr))
+  col_params <- t(Q)
+  Qty <- t(as.matrix(mat %*% Q))  ## row_params = Qty
+  wrapMatrix(
+    "TransformLinearResidual", 
+    convert_matrix_type(mat, "double"),
+    row_params = Qty,
+    col_params = col_params,
+    global_params = numeric()
+  )
```

Comment:
You should add in a check for `storage_order(mat)` here, and if it's `"row"`, then I think you'll want to swap the row_params and col_params to make your code work out. (The C++ layer thinks of everything as colum-major, so an R matrix with `storage_order(mat) == row` will be seen by C++ as `t(mat)`. Luckily the math of this transform is nearly the same in C++ for `t(mat)`)